File: android-bridge/app/src/main/java/com/jaato/a11ybridge/transport/BinaryFrame.kt
@file:OptIn(ExperimentalSerializationApi::class)

package com.jaato.a11ybridge.transport

import kotlinx.serialization.EncodeDefault
import kotlinx.serialization.ExperimentalSerializationApi
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import okio.Buffer
import okio.ByteString

/**
 * Self-describing binary frame (PROTOCOL §4):
 *
 *   [ 4-byte BE uint32 headerLen ][ headerLen bytes UTF-8 JSON header ][ payload bytes ]
 *
 * No base64, no ordering assumptions — correlation is explicit in the header.
 */
object BinaryFrame {

    fun frame(headerJson: String, payload: ByteArray): ByteString {
        val header = headerJson.encodeToByteArray()
        val buf = Buffer()
        buf.writeInt(header.size)   // Okio writeInt is big-endian
        buf.write(header)
        buf.write(payload)
        return buf.readByteString()
    }
}

/**
 * Header schema for binary frames (§4). `snapshotVersion` is present only when the blob
 * is tied to a tree version (bundled captures / settle); `reason` ∈ on_demand | bundled.
 */
@Serializable
data class ShotHeader(
    @EncodeDefault(EncodeDefault.Mode.ALWAYS) val type: String = "screenshot",
    val correlationId: String,
    val snapshotVersion: Long? = null,
    val format: String,
    val width: Int,
    val height: Int,
    val reason: String,
) {
    fun toJson(): String = Wire.json.encodeToString(this)

    companion object {
        const val REASON_ON_DEMAND = "on_demand"
        const val REASON_BUNDLED = "bundled"
    }
}

File: windows-bridge/JaatoBridge/Transport/BinaryFrame.cs
using System.Buffers.Binary;
using System.Text;
using System.Text.Json;

namespace JaatoBridge.Transport;

/// <summary>
/// §4 binary framing: [4-byte BE headerLen][UTF-8 JSON header][raw blob].
/// The device only ever <em>builds</em> these (screenshots); daemon→device is text-only.
/// </summary>
public static class BinaryFrame
{
    public static byte[] Build(object header, ReadOnlySpan<byte> payload)
    {
        byte[] headerBytes = JsonSerializer.SerializeToUtf8Bytes(header, Wire.Json);
        var frame = new byte[4 + headerBytes.Length + payload.Length];
        BinaryPrimitives.WriteUInt32BigEndian(frame.AsSpan(0, 4), (uint)headerBytes.Length);
        headerBytes.CopyTo(frame.AsSpan(4));
        payload.CopyTo(frame.AsSpan(4 + headerBytes.Length));
        return frame;
    }
}

/// <summary>Header schema for a screenshot binary frame (§4).</summary>
public sealed class ScreenshotHeader
{
    public string Type => "screenshot";
    public string CorrelationId { get; init; } = "";
    public long SnapshotVersion { get; init; }
    public string Format { get; init; } = "webp";
    public int Width { get; init; }
    public int Height { get; init; }
    public string Reason { get; init; } = "on_demand"; // on_demand | bundled
}
